cmd/eskip/readclient.go

```diff
@@ -0,0 +1,127 @@
+package main
+
+import (
+	"github.com/zalando/skipper/eskip"
+	"github.com/zalando/skipper/eskipfile"
+	etcdclient "github.com/zalando/skipper/etcd"
+	innkeeperclient "github.com/zalando/skipper/innkeeper"
+	"io"
+	"io/ioutil"
+	"os"
+)
+
+type ReadClient interface {
+	LoadAndParseAll() ([]*eskip.RouteInfo, error)
+
+	LoadUpdate() (eskip.RouteList, []string, error)
+}
+
+type stdinReader struct {
+	reader io.Reader
+}
+
+type inlineReader struct {
+	routes string
+}
+
+type idsReader struct {
+	ids []string
+}
+
+// TODO eliminate duplicate code in write and read clients
+func createReadClient(m *medium) (ReadClient, error) {
```

Comment:
I decoupled the place where we create the stateless object from the place where we use them in order to follow the [Dependency Inversion Principle](https://en.wikipedia.org/wiki/Dependency_inversion_principle). Separation so concerns is the second principle I'd like to follow. It is really important when building complex systems to decouple concerns. Looking at our example, after separating the client creation concern and the loading concern: the readability is improved, as the load.go file is not much smaller. Also, as soon as you want to add a new reader, you don't have to understand what everything happens in the load file. You just create a new reader, in the separate file. Testability is improved, as you are now able to mock the ReadClient.

The [Single Responsibility Principle](https://en.wikipedia.org/wiki/Single_responsibility_principle) is the second princile I'd like to respect. And abstraction which needs to load routes, should only do one thing load the routes, as it's name suggests. Creating a client is another concern, and it's the responsibility of another abstraction.

Regarding your comments:
- yes, the switch is still there, but you don't have to dig inside of the load.go file to find it. As soon as you need to deal with a readclient, you open its file and add, or modify a client.
- yes, the code has a few more lines. But now it's consistent. Before the change, we were many kinds of abstractions in order to read the lines. For innkeeper, etcd and file we had the DataClient interface, and for the rest we had inline implementations. Now, when we want to read  routes, we have one abstraction: ReadClient. And having less lines of code doesn't make the code more understandable (see bad written Scala :P)